Add experimental musl support

yuwata opened the pull request:
This adds experimental support of building systemd with musl. This allows to build systemd with musl and run most of the unit tests. Running integration tests is not yet supported, however. Hopefully, this should be a good starting point to support musl.

This requires musl-1.2.5 with https://github.com/kraj/musl/commit/fde29c04adbab9d5b081bf6717b5458188647f1c. The patch is already backported to Alpine/postmarketOS's musl package since musl-1.2.5-r11. See https://gitlab.alpinelinux.org/alpine/aports/-/commit/333e50c2052c74bb372f821fdeea2e6814f2fe7a.

yuwata committed:
musl: glob-util: filter out . and .. even if GLOB_ALTDIRFUNC is not supported

musl neither support GLOB_ALTDIRFUNC nor GLOB_BRACE.
Let's make safe_glob() work even when GLOB_ALTDIRFUNC is not supported.
Currently, GLOB_BRACE is simply ignored when it is not supported.

esmil committed:
musl: add fallback parse_printf_format() implementation

musl does not provide parse_printf_format(). Let's introduce a fallback
method.

Co-authored-by: Yu Watanabe <watanabe.yu+github@gmail.com>

yuwata committed:
musl: ci: add build test and unit tests

poettering commented on meson.build:
```diff
@@ -72,7 +72,10 @@ conf.set10('SD_BOOT', false)
 
 # Create a title-less summary section early, so it ends up first in the output.
 # More items are added later after they have been detected.
-summary({'build mode' : get_option('mode')})
```
in the commit msg: skelton → skeleton

poettering commented on meson.build:
```diff
@@ -1004,6 +1004,23 @@ librt = cc.find_library('rt')
 
+# On some distributions that uses musl (e.g. Alpine), libintl.h may be provided by gettext rather than musl.
+# In that case, we need to explicitly link with libintl.so.
+if get_option('libc') == 'musl'
```
why condition this check on musl? if this check runs on glibc it should come to the correct reply, not?

I'd always avoid special casing libc's wherever we can, and check for features, not check for implementations.

yuwata replied:
~~This is for build test, in that case both musl and glibc with libxcrypt are installed, and we should not use libxcrypt in that case.
But, that's really artificial setup. I will drop the change, and workaround the build test.~~

poettering commented on meson.build:
```diff
@@ -1015,10 +1015,14 @@ else
+        if not libcrypt.found()
+                # fallback to use find_library() if libcrypt is provided by glibc, e.g. for LibreELEC.
+                libcrypt = cc.find_library('crypt')
+        endif
```
these days i think i'd prefer if we alwys use libxcrypt, even on systems where libc provides it. i mean, libxcrypt should build fine on musl, no? hence we could use it there, too?

yuwata replied:
Yeah, waiting in #38974.

But, as commented at https://github.com/systemd/systemd/issues/39413#issuecomment-3442100342, libxcrypt cannot be built musl. Hence we cannot use it when building with musl. But that's another topic here. We can introduce tiny wrapper for musl after #38974.

rossburton replied:
From a few minutes of poking, libxcrypt builds fine against musl. The big problem is that musl and libxcrypt both have `/usr/include/crypt.h`, with the latter being a superset.

Tempting to see if upstream musl would be interested in support for disabling crypt so people who want xcrypt can use it.

poettering commented on meson.build:
```diff
@@ -560,6 +560,12 @@ conf.set10('HAVE_WARNING_ZERO_AS_NULL_POINTER_CONSTANT', have)
+        # glibc always defines _LARGEFILE64_SOURCE when _GNU_SOURCE is set, but musl does not do that,
+        # and it is necessary for making getdents64() and struct dirent64 exist.
+        conf.set('_LARGEFILE64_SOURCE', 1)
+endif
```
why condition this? seems fine to always set this, no?

yuwata replied:
That's totally fine, will drop the condition.

poettering commented on src/boot/meson.build:
```diff
@@ -182,6 +182,11 @@ if get_option('mode') == 'developer' and get_option('debug')
 
+if get_option('libc') == 'musl'
+        # To make musl not define wchar_t as int, rather than short.
+        efi_c_args += '-D__DEFINED_wchar_t'
```
i don't grok this? this is for efi code, which doesn't use either libc?

yuwata replied:
~~Indeed. Maybe we should not use musl-gcc for building sd-boot. I will check.~~

Even libc is not linked, very basic headers by libc (both glibc and musl) are included. Hence, this is necessary.

poettering commented on meson.build:
```diff
@@ -581,6 +581,7 @@ assert(long_max > 100000)
 conf.set_quoted('LONG_MAX_STR', f'@long_max@')
 
 foreach ident : [
+        ['renameat2',         '''#include <stdio.h>''',         get_option('libc') == 'musl'], # since musl-1.2.6
```
why condition on musl? check should work fine on glibc too

yuwata replied:
Totally fine to drop the condition. Will drop.

poettering commented on meson.build:
```diff
@@ -1678,7 +1674,6 @@ foreach term : ['analyze',
                 'tmpfiles',
                 'tpm',
                 'userdb',
-                'utmp',
```
i thought people use "utmps" for this on musl?

yuwata replied:
What do you mean? This is our meson option name in meson_options.txt.

poettering replied:
"utmps" seems to be a polyfill lib for utmp for musl systems

yuwata replied:
Interesting. I did not know that. Will check.

poettering commented on meson.build:
```diff
@@ -1688,14 +1683,34 @@ endforeach
+                'utmp']
+
+        have = get_option(term)
+        if have and get_option('libc') == 'musl'
```
as above, why not make these check for features? i.e. if gshadow.h is available, if NI_IDN is available, if nss.h is available?

yuwata replied:
OK will do that.

poettering commented on src/include/musl/printf.h:
```diff
@@ -0,0 +1,35 @@
+#include <features.h>
+#include <stddef.h>
+
+#ifdef __GLIBC__
```
shouldnt this check for existance of parse_printf_format, i.e. HAVE_PARSE_PRINTF_FORMAT?

poettering commented on src/libc/musl/printf.c:
```diff
@@ -0,0 +1,276 @@
+        for (i = 0; i < last; i++)
+                types[i] = state_to_pa_type(types[i]);
+
+        return last;
```
grmbl, i really dislike that this reimplements printf parsing here. it sucks really hard that musl doesn't support that natively

yuwata replied:
Yeah, but there is nothing we can do here.

poettering commented on src/libc/musl/string.c:
```diff
@@ -0,0 +1,26 @@
+        } else {
+                memcpy(buf, msg, l + 1);
+                errno = saved_errno;
+        }
```
hmm, i don't grok this one.

first of all, i think this GNU wrapper should wrap the POSIX strerror_r(), not simple strerror(). Moreover, GNU's version *may* use the provided buffer, but doesn't have to, as per docs, hence no need to unconditionally copy this here...

poettering commented on src/libc/musl/time.c:
```diff
@@ -0,0 +1,86 @@
+        if (!s || !format || !tm)
+                return NULL;
+
+        /* We only fallback if the format is exactly "%z". */
```
umpf, this is really magic, checking for one specific format string only

yuwata replied:
Yeah, but we only use `%z` sorely, not combined with other format specifier. I do not think we should provide complete version.

poettering commented on src/libc/musl/generate-strerrorname_np.sh:
```diff
@@ -0,0 +1,39 @@
+        if ((size_t) errnum >= sizeof(errno_table) / sizeof(errno_table[0]))
+                return NULL;
+        return errno_table[errnum];
+}
```
this seems entirely redundant? src/basic/errno-list.c already deals with strerrorname_np() being absent

yuwata replied:
Right, at least now. But we will soon bump the requirements for glibc, and the fallback logic will be dropped then.

poettering replied:
well, if you want to keep musl around, we can't really drop it...

also, the table is there anyway, since glibc has no api to translate error strings back to numbers, so we need to keep the table anyway.

poettering commented on src/include/musl/features.h:
```diff
@@ -0,0 +1,8 @@
+#include_next <features.h>
+
+#ifndef __THROW
+#define __THROW
```
hmm __THROW is a c++ compat thing, we don#t need that. I think the right approach would be to just drop it from our compat headers.

poettering replied:
(it's not just a c++ thing, it's also a glibc internal thing... we really should avoid use of internal glibc stuff)

yuwata replied:
@DaanDeMeyer WDYT about dropping `__THROW` from our compat headers? I really have no idea what the attribute does...

poettering replied:
it sets the throw gcc attribute, which is c++ stuff, not relevant to us

https://stackoverflow.com/questions/24882645/gcc-what-is-attribute-nothrow-used-for

yuwata replied:
OK. Then let's drop them.

poettering commented on src/include/musl/sys/prctl.h:
```diff
@@ -0,0 +1,9 @@
+#include <features.h>
+#include <linux/prctl.h>        /* IWYU pragma: export */
+
+int prctl(int, ...);
```
hmm, i dont grok this? what kind of conflicts?

yuwata replied:
musl's sys/prctl.h and linux/prctl.h both provides `struct prctl_mm_map` thus they conflict with each other.

At several places in our code, both sys/prctl.h and linux/prctl.h are included...

poettering replied:
hmm, can't we instead just include one of the two? do we really need to include both on glibc?

poettering commented on src/include/musl/linux/if_ether.h:
```diff
@@ -0,0 +1,5 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
+
+#undef __UAPI_DEF_ETHHDR
```
please add comment here, what this is for

yuwata replied:
It is not necessary anymore. I forgot to drop.

poettering commented on src/include/musl/fcntl.h:
```diff
@@ -0,0 +1,14 @@
+ * forward.h. To avoid the conflict, here temporary undef AT_FDCWD before including fcntl.h. */
+#ifdef AT_FDCWD
+#undef AT_FDCWD
+#endif
```
the commit msg should say basic-forward.h, not forward.h

shiuldn't we instead drop AT_FDCWD from basic-forward.h? @DaanDeMeyer?

yuwata replied:
I think we should keep it. As many headers use AT_FDCWD. Rather than complicated undef hack, I conditionalize the definition in basic-forward.h.

poettering commented on src/include/musl/limits.h:
```diff
@@ -0,0 +1,8 @@
+
+/* HOST_NAME_MAX should be 64 on linux, but musl uses the one by POSIX (255). */
+#undef HOST_NAME_MAX
+#define HOST_NAME_MAX 64
```
this is so broken in musl, because the kernel only accepts 64 bytes...

i think it would be safer, to maybe introduce a new macro

```c
#define LINUX_HOST_NAME_MAX CONST_MIN(HOST_NAME_MAX, 64U)
```

and then use that everywhere.

poettering replied:
i.e. add that to hostname-util.h

poettering replied:
i mean, i am fine with simply polyfills for musl. But changing musl definitions to something else seems problematic

rossburton replied:
Feels like this should be fixed in musl instead? If Linux has a hard limit of 64 bytes and musl targets Linux exclusively, then setting it to 255 is just wrong.

poettering replied:
@rossburton musl guy doesn't think reality or linux matters:

https://www.openwall.com/lists/musl/2019/02/14/2

poettering commented on src/include/musl/sched.h:
```diff
@@ -0,0 +1,30 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+#pragma once
```
i'd really prefer if we'd just change the one line in our sources where this matters, i.e. move the --i one line before... 

i really dislike if() tests with side-effects anyway, i think it would be nicer to read if split up.

i also think that musl is simplly buggy here. is there a bug file about this?

poettering commented on src/basic/time-util.c:
```diff
@@ -921,8 +925,11 @@ static int parse_timestamp_impl(
         assert(minus == 0);
 
+#ifdef __GLIBC__
+        /* musl does not set tm_wday field and set 0 unless it is explicitly requested by %w or so. */
```
hmm, so musl claims every date way sunday? yuck... smells like a bug...

sounds fine to not fill this in, i.e. setting this to -1, but setting it to 0 (i.e. sunday) is just borked...

can you file a bug against musl?

poettering commented on src/basic/glob-util.c:
```diff
@@ -23,15 +56,53 @@ int safe_glob_full(const char *path, int flags, opendir_t opendir_func, char ***
+#if GLOB_ALTDIRFUNC == 0
+        use_gnu_extension = false;
+#else
+        SET_FLAG(flags, GLOB_ALTDIRFUNC, use_gnu_extension);
```
hmm, not supporting GLOB_ALTDIRFUNC is kinda problematic. This stuff is used by tmpfiles, and globbing there shouldn't touch atime of those dirs, but on musl it will if you ignore this here. this means that use of globs will neuter the time-based clean-up stuff...

yuwata replied:
Yeah, but there is nothing we can do...

poettering replied:
well, that suggests we shouldn't use glob() on musl at all?

poettering commented on src/test/test-errno-util.c:
```diff
@@ -26,18 +26,28 @@ TEST(STRERROR) {
+#else
+        /* musl provides catch all error message for unknown error number. */
+        ASSERT_STREQ(a, "No error information");
+        ASSERT_STREQ(b, "No error information");
```
hmm, i am not a fan of this, i'd really prefer if we'd expose the same behaviour everywhere.

hence, maybe introduce strerror_safe() or so, which hides the musl and glibc differences away, and then use that everywhere.

yuwata replied:
Error strings by musl and glibc are anyway different...

poettering replied:
i think it's ok if the error strings differ a little, but "No error information" is not just a little bit different. i.e. unknown errorr numbers are not "no error information", they very much are, hence I think the abstraction would be useful

poettering commented on src/test/test-fileio.c:
```diff
@@ -400,7 +400,15 @@ TEST(write_string_stream) {
+        ASSERT_ERROR(write_string_stream(f, "boohoo", 0), EBADF);
+#else
+        /* Even the file is opened with the read-only mode, fputs() and fputc() by musl succeed but nothing
+         * actually written, thus write_string_stream() also succeeds. */
```
hmm, what? that's super broken. please file a bug.

would prefer if write_string_stream() would generate EBADF in this case somehow. can we detect in which way a FILE is open?

yuwata replied:
Maybe? But that's for testing?

poettering replied:
i'd really prefer if write_string_stream() would fail the same everywhere...

poettering commented on src/test/test-locale-util.c:
```diff
@@ -50,7 +50,12 @@ TEST(locale_is_installed) {
+#ifdef __GLIBC__
+        ASSERT_OK_ZERO(locale_is_installed("zz_ZZ"));
+#else
+        /* musl seems to return a non-null locale object even if it is not installed. */
```
hmm, this is really broken too, any chance you can file a bug?

is there a way to detect these kinds of fake locales, and refuse them?

poettering commented on src/test/test-seccomp.c:
```diff
@@ -557,9 +557,13 @@ TEST(memory_deny_write_execute_mmap) {
                 assert_se(p == MAP_FAILED);
-                assert_se(errno == EPERM);
+#  ifdef __GLIBC__
+                ASSERT_EQ(errno, EPERM);
```
hmm, this doesn't look right. This means the seccomp filter doesn't work, and needs to be fixed!

what does glibc's mmap() do differently than musl's?

or is musl manually patching around the error codes? that'd suck

yuwata replied:
musl maps EPERM to ENOMEM. See https://github.com/kraj/musl/blob/0784374d561435f7c787a555aeab8ede699ed298/src/mman/mmap.c#L33-L35

poettering commented on src/test/test-time-util.c:
```diff
@@ -400,6 +400,25 @@ static void test_format_timestamp_impl(usec_t x) {
 
+#ifndef __GLIBC__
+        /* Because of the timezone change, format_timestamp() may set timezone that is currently unused.
+         * E.g. Africa/Juba may set EAT, but currently it uses CAT/CAST. */
```
i don't grok this?

yuwata replied:
I will extend the comment.

poettering commented on src/test/test-time-util.c:
```diff
@@ -1110,7 +1129,12 @@ TEST(in_utc_timezone) {
         ASSERT_STREQ(tzname[1], "UTC");
+#else
+        /* musl sets an empty string to tzname[1] when DST is not used by the timezone. */
+        ASSERT_STREQ(tzname[1], "");
```
i'd not do the special casing here, and instead just allow both isempty() and "UTC"

poettering replied:
or maybe add a helper: 

```c
static const char *get_tzname(bool dst) {

        /* musl leaves the DST timezone name unset if there is no DST, map this back to no DST */
        if (dst && isempty(tzname[1]))
                dst = false;
   
        if (isempty(tname[dst]))
                return NULL;

        return tzname[dst]; 
}
```

and then use that everywhere, i.e. in parse_timestamp() too
or so

poettering commented on meson.build:
```diff
@@ -572,6 +572,13 @@ conf.set('SIZEOF_RLIM_T', cc.sizeof('rlim_t', prefix : '#include <sys/resource.h
+                error('Unexpected size of rlim_t: @0@'.format(conf.get('SIZEOF_RLIM_T')))
+        endif
+        conf.set_quoted('RLIM_FMT', '%llu')
+endif
```
why is this implemented in meson here? might as well implement in macro alone in format-util.h no?

poettering commented on meson.build:
```diff
@@ -1610,11 +1652,9 @@ foreach term : ['analyze',
                 'hibernate',
                 'hostnamed',
                 'hwdb',
-                'idn',
```
hmm, i thought idn was implemented via libidn2, which should work fine on musl?

yuwata replied:
Not sure. At least NI_IDN is not provided.

poettering replied:
ah, right, that's a different thing indeed. ENABLE_IDN and HAVE_LIBIDN2 are two distinct knobs..

keszybz commented on .github/workflows/unit-tests-musl.sh:
```diff
@@ -0,0 +1,59 @@
+            info "Run phase"
+
+            # Create dummy machine ID.
+            echo '052e58f661f94bd080e258b96aea3f7b' > /etc/machine-id
```
No space after `>` please.

yuwata committed:
musl: meson: add libutmps support

musl only provides fake functions, hence even with -Dutmp=true,
utmp/wtmp functionalities do not work. Let's allow to build with
libutmps.

yuwata committed:
musl: utmpx: add several missing definitions

keszybz reviewed (approved):
I'll go ahead and merge this. It all does seem fairly fragile, so I'm a bit worried that maintenance will be a problem. But the only way to verify this is to merge this and let people play with it.

I spent some time repeating the build with musl locally on Fedora, and that part works well enough. But I couldn't get the tests to run without a container. That part is a bit annoying — we really need to have a simple local workflow for development. But at least compilation works, so maybe it won't be so bad.

Compilation and unit tests pass in the CI, which is good enough for now.
